File: LeetCodes/binary-subarrays-with-sum/12-10-2023, 5_09_34 PM/binary-subarrays-with-sum.cpp
// Link: https://leetcode.com/problems/binary-subarrays-with-sum

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: binary-subarrays-with-sum
 *  RUNTIME: 1204 ms
 *  MEMORY: 29.1 MB
 *  DATE: 12-10-2023, 5:09:34 PM
 *
 */

// Solution:

class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int left=0;
        int right=0;
        int count=0;
        int sum=0;
        while(right<nums.size())
        {
            sum+=nums[right];
            while(left<right && sum>goal)
            {
                sum-=nums[left];
                left+=1;
            }
            if(sum==goal)
            {
                int temp=left;
                while(temp<right && nums[temp]==0)
                {
                    temp+=1;
                    count+=1;
                }
                count+=1;
            }
            right+=1;
        }
        return count;
    }
};

File: LeetCodes/binary-subarrays-with-sum/3-14-2024, 10_01_31 AM/binary-subarrays-with-sum.java
// Link: https://leetcode.com/problems/binary-subarrays-with-sum

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: binary-subarrays-with-sum
 *  RUNTIME: 3 ms
 *  MEMORY: 48.6 MB
 *  DATE: 3-14-2024, 10:01:31 AM
 *
 */

// Solution:

class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int totalCount = 0;
        int sum = 0;
        int prefixSum[] = new int[nums.length + 1];
        prefixSum[0] = 1;
        
        for (int num : nums) {
            sum += num;
            if (sum >= goal) {
                totalCount += prefixSum[sum - goal];
            }
            prefixSum[sum]++;
        }
        
        return totalCount;
    }
}
